ubenchmark/src/main/java/org/apache/calcite/benchmarks/StringConstructBenchmark.java

```diff
@@ -0,0 +1,133 @@
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.infra.Blackhole;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.io.Writer;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A benchmark of the most common patterns that are used to construct gradually String objects.
+ *
+ * The benchmark emphasizes on the build patterns that appear in the Calcite project.
+ */
+@Fork(value = 1, jvmArgsPrepend = "-Xmx2048m")
+@Measurement(iterations = 10, time = 100, timeUnit = TimeUnit.MILLISECONDS)
+@Warmup(iterations = 10, time = 100, timeUnit = TimeUnit.MILLISECONDS)
+@Threads(1)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@BenchmarkMode(Mode.Throughput)
+public class StringConstructBenchmark {
+
+  /**
+   * A state holding a Writer object which is initialized only once at the beginning of the
+   * benchmark.
+   */
+  @State(Scope.Thread)
+  public static class WriterState {
+    public Writer writer;
+
+    @Setup(Level.Trial)
+    public void setup() {
+      this.writer = new StringWriter();
+    }
+  }
+
+  /**
+   * A state holding an Appendable object which is initialized once at the beginning of every
+   * iteration.
+   */
+  @State(Scope.Thread)
+  public static class AppenderState {
+    /**
+     * The type of the appender to be initialised.
+     */
+    @Param({"StringBuilder", "StringWriter", "PrintWriter"})
+    public String appenderType;
+
+    public Appendable appender;
+
+    @Setup(Level.Iteration)
+    public void setup() {
+      if (appenderType.equals("StringBuilder")) {
+        this.appender = new StringBuilder();
+      }
+      if (appenderType.equals("StringWriter")) {
+        this.appender = new StringWriter();
+      }
+      if (appenderType.equals("PrintWriter")) {
+        this.appender = new PrintWriter(new StringWriter());
+      }
+      throw new IllegalStateException("No correct appender type is specified.");
+    }
+  }
+
+  @Benchmark
+  public StringBuilder initStringBuilder() {
+    return new StringBuilder();
+  }
+
+  @Benchmark
+  public StringWriter initStringWriter() {
+    return new StringWriter();
+  }
+
+  @Benchmark
+  public PrintWriter initPrintWriter(WriterState writerState) {
+    return new PrintWriter(writerState.writer);
+  }
+
+
+  @Benchmark
+  public void appendString(Blackhole bh, AppenderState appenderState) throws IOException {
+    bh.consume(appenderState.appender.append("A small string"));
```

Comment:
I'm afraid this is wrong.

In fact you want to compare benchmarks that have a steady state.
However the more you append the slower it becomes.

You might want to add `@Param int numAppends;` and compare perf for the string of a given length.

Alternative option is to call RelLiteral#toJavaString and run with different Calcite versions in the classpath.

Reply:
Yes, I was aware of that but I think the comparison is fair because all appenders are initialized at the beginning of each operation. However, it is true that I am not just measuring the effect of a single append operation but also the effect of resize.

Another alternative option is to measure exactly the pattern which appears in RexLiteral#toJavaString which is 
`StringWriter sw = new StringWriter();
PrintWriter pw = new PrintWriter(sw);
pw.print("String");
pw.flush();
sw.toString()`

vs

`StringBuilder sb = new StringBuilder();
sb.append("String");
sb.toString()` 

By the way this is what I started with so I already know the answer :)

Reply:
> exactly the pattern which appears in RexLiteral#toJavaString which is

That's ok.

If only we could use `equals/hashCode` and/or kind of `Hashable` vs `EqualsCapable` interfaces...

Koloboke seems to be a nice `map` library which supports custom equals/hashcode implementations: https://github.com/leventov/Koloboke/blob/302e6d5e5889a80edb3aeed2e664f480439f8155/compile/tutorial.md#custom-key-equivalence

Reply:
By the way: I have seen that `PrintWriter`, however I thought "it was decent enough", and I did not feel brave enough to break public API.

Reply:
I didn't know about Koloboke. Thanks for sharing! 

> By the way: I have seen that PrintWriter, however I thought "it was decent enough", and I did not feel brave enough to break public API.

I was thinking to try to get rid off PrintWriter from everywhere but this would introduce too many changes and many of them unjustifiable. However, this piece of code is important to be fast and it makes sense to improve it a bit, especially if the change needed is so small.

Reply:
@vlsi : I changed a bit the `appendString` benchmark to include `@Param int maxAppends` so now it also includes the time to instantiate the Appendable. This is a more realistic representation of what is happening in `RexLiteral#toJavaString`.